Impliment PlayerAPI

davidbrowning opened the pull request:
I think this is how the callbacks are going to work. Let me know if something needs to be changed. Still need to implement the update scoring functions.

davidbrowning committed:
HouseKeeping

davidbrowning committed:
More Tidying, Shortening functions

davidbrowning committed:
Wrote a shorter test and added an empty spades constructor for convenience

davidbrowning committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into michaelkamerath/master

davidbrowning committed:
integrate w/player api

davidbrowning committed:
merge issues

davidbrowning committed:
forgot a semi colon

davidbrowning committed:
fixed test

davidbrowning committed:
Derp moments

Fixed the slashes and that for loop was only supposed to go to 4

C-Merrill committed:
clang format

davidbrowning committed:
Writing in callbacks to playerAPI

davidbrowning committed:
merge conflicts

davidbrowning committed:
missed some of the merge issues

katiesweet commented on source/GameLogic/SpadesLogic.cpp:
```diff
@@ -87,15 +92,22 @@ Spades::Spades(std::vector<std::shared_ptr<Player>> p)
-  for (auto&& p : players)
+	  for (auto&& p : players)
   {
     p->setValidateBid([this](int b) { receiveBid(b); });
```
Not sure you need this line since you have it in the constructor...

davidbrowning replied:
You're totally right, changing right now.

C-Merrill committed:
clang format

C-Merrill commented:
@katiesweet can you verify the functionality here?
@davidbrowning let's get going on those unit tests

katiesweet reviewed (commented):
I think the way you're doing callbacks should work... It was not quite the way I had in mind to do this, but I don't see why it wouldn't...

davidbrowning committed:
Missed part of the scoring logic

specifically adjusting the bags at the end of the round

davidbrowning committed:
removed redundant setValidateBid

davidbrowning committed:
Re-Work Spades Logic

Should now be asyncronous and work with the player API

katiesweet commented on source/GameLogic/SpadesLogic.cpp:
```diff
@@ -73,30 +73,76 @@ int next(int plId)
+    trickWinner = getTrickWinner(trick, trickWinner);
+    if (players.at(turn)->getHand().size() < 1)
+    {
+      s = ROUND_OVER;
```
This isn't totally necessary, but it could be nice to extract this section into a roundOver() function

katiesweet commented on source/GameLogic/SpadesLogic.cpp:
```diff
@@ -73,30 +73,76 @@ int next(int plId)
+      score();
+      UpdateGameStateMessage();
+      setDeck();
+      for (auto player : players)
```
Maybe a startNewRound() function as well... again, not necessary, but a thought. 

Also, when starting a new round, you need to call player.startNewRound() to manage the updating of players scores, reseting round score, reseting their bid, etc.

davidbrowning commented:
Based on what we did in class, I'm thinking this how the logic is supposed to work.

katiesweet commented on source/GameLogic/SpadesLogic.hpp:
```diff
@@ -2,31 +2,36 @@
   bool spadesBroken;
   Card globalCard;
+  std::atomic<int> bidPlaceHolder;
+  Card movePlaceHolder;
```
You might be able to get rid of these bidPlaceHolder/morePlaceHolder since we went a different route now.

katiesweet commented on source/GameLogic/SpadesLogic.cpp:
```diff
@@ -73,30 +73,76 @@ int next(int plId)
 void Spades::receiveValidMove(Card c)
 {
-  // the return card is coming here!!
+  trick.push_back(c);
```
You could wait to do this until after you know it is valid... Also, when are your removing this card from the players hand?

davidbrowning replied:
I suppose so, but I felt it was easier to get the card to the validMove function this way and then I only have to change it when an invalid move is played (which is hopefully a pretty small number of times if we really are playing with 4 humans).

katiesweet reviewed (changes requested):
@davidbrowning Overall, this looks a ton better! Good work! Let me know if you have questions about any of my comments. 

Two more things... 1. I'm a bit concerned about your getNextPlayer() function. It only increments plId, never sets it back to 0, so it'll call turn 0 over and over again once it hits plId = 3. 

Also, there is a lot of legacy code in here from other versions of your game. It would help readability quite a bit if it were cleaned up a little.

davidbrowning committed:
Update SpadesLogic.cpp

added startNewRound() and roundOver(). Also added the player->startNewRound() function where applicable.

davidbrowning committed:
added the new functions

Also removed some leftover stuff

davidbrowning commented:
getNextPlayer() does return 0 whenever it receives 3 (or if something terrible happened something higher than 3).

katiesweet commented on source/GameLogic/SpadesLogic.cpp:
```diff
@@ -71,32 +71,89 @@ int next(int plId)
 void Spades::receiveValidMove(Card c)
 {
-  // the return card is coming here!!
+  trick.push_back(c);
```
This line isn't necessarily the issue...
What I think is the problem is that you are adding the card to the trick, but you are never removing that card from the players hand. Down below, you are inserting it back into the hand if it's an invalid move, but it was never removed from the hand in the first place, so you'd have two of the same card in the hand...

davidbrowning replied:
Ah now I see. I'll get right on it

davidbrowning replied:
So it should just be that line that fixes it right?

 +  players.at(turn)->removeCardFromHand(c);

katiesweet reviewed (changes requested):
@davidbrowning This is looking good! Sorry about the getNextPlayer function. I misinterpreted what was going on. However, I still think we have an issue where we are never removing the card from the players hand after they make a move.

davidbrowning committed:
removeCardFromHand(c) added

katiesweet commented:
Yep, I think that should do it. Thanks @davidbrowning! 

@C-Merrill and @michaelkamerath I think this one is about ready to go. Will you please look over it quickly?

katiesweet commented on source/GameLogic/SpadesLogic.cpp:
```diff
@@ -71,32 +71,90 @@ int next(int plId)
+  }
+  if (trick.size() < 4)
+  {
+    players.at(turn)->requestMove();
```
You should probably UpdateGameStateMessage() before this line or after you've determined it is a valid move. This is because once you've played a card onto the trick, everyone is going to want to update their GUI's to reflect that.

katiesweet commented on source/GameLogic/SpadesLogic.cpp:
```diff
@@ -71,32 +71,90 @@ int next(int plId)
+    {
+      turn = trickWinner;
+      players.at(turn)->incrementTricksWon();
+      players.at(turn)->requestMove();
```
Same here. It would be nice to have an UpdateGameStateMessage() above this requestMove since you generally want to update your GUI after someone finishes their turn/you cleared the field since a trick is over.

katiesweet reviewed (changes requested):
Nevermind, sorry guys, one more quick thing @davidbrowning. You aren't calling UpdateGameStateMessage() besides at roundOver/gameOver. This is the function that updates everyone's GUI's with what is going on. Therefore, after every turn is over (before requesting a move from someone else), pleas call UpdateGameStateMessage() to alert everyone of the card that has been played.

davidbrowning commented:
Cool! I wasn't sure what exactly that one did so I'll put that in as soon
as I can.

On Dec 2, 2016 11:16 AM, "katiesweet" <notifications@github.com> wrote:

> *@katiesweet* requested changes on this pull request.
>
> Nevermind, sorry guys, one more quick thing @davidbrowning
> <https://github.com/davidbrowning>. You aren't calling
> UpdateGameStateMessage() besides at roundOver/gameOver. This is the
> function that updates everyone's GUI's with what is going on. Therefore,
> after every turn is over (before requesting a move from someone else),
> pleas call UpdateGameStateMessage() to alert everyone of the card that has
> been played.
> ------------------------------
>
> In source/GameLogic/SpadesLogic.cpp
> <https://github.com/michaelkamerath/SharedServer/pull/70#pullrequestreview-11211210>
> :
>
> > -  // the return card is coming here!!
> +  players.at(turn)->removeCardFromHand(c);
> +  trick.push_back(c);
> +  if (validMove())
> +  {
> +    turn = getNextPlayer(turn);
> +  }
> +  else
> +  {
> +    auto c = trick.back();
> +    players.at(turn)->insertCardToHand(c);
> +    trick.pop_back();
> +  }
> +  if (trick.size() < 4)
> +  {
> +    players.at(turn)->requestMove();
>
> You should probably UpdateGameStateMessage() before this line or after
> you've determined it is a valid move. This is because once you've played a
> card onto the trick, everyone is going to want to update their GUI's to
> reflect that.
> ------------------------------
>
> In source/GameLogic/SpadesLogic.cpp
> <https://github.com/michaelkamerath/SharedServer/pull/70#pullrequestreview-11211210>
> :
>
> > +  {
> +    players.at(turn)->requestMove();
> +  }
> +  else
> +  {
> +    trickWinner = getTrickWinner(trick, trickWinner);
> +    if (players.at(turn)->getHand().size() < 1)
> +    {
> +     roundOver();
> +     startNewRound();
> +    }
> +    else
> +    {
> +      turn = trickWinner;
> +      players.at(turn)->incrementTricksWon();
> +      players.at(turn)->requestMove();
>
> Same here. It would be nice to have an UpdateGameStateMessage() above this
> requestMove since you generally want to update your GUI after someone
> finishes their turn/you cleared the field since a trick is over.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/michaelkamerath/SharedServer/pull/70#pullrequestreview-11211210>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APW1YhXU7aBJdXiR_t9XA1ZrXQSF1uQVks5rEGB3gaJpZM4LA81j>
> .
>

C-Merrill commented:
@katiesweet check linux build once that change comes in and I think we can merge this once that's verified.

davidbrowning committed:
UpdateGameStateMessage() additions

davidbrowning commented:
I don't have the linux build working on my machine. @adamtew can you pull and build it?

davidbrowning committed:
merge conflicts

katiesweet commented:
Builds fine for me! I think this one is ready @davidbrowning @C-Merrill @michaelkamerath !

C-Merrill committed:
Merge remote-tracking branch 'origin/master' into davidbrowning

C-Merrill commented:
I merged master in to make sure memory leaks while running tests were those that were already solve in Shem's merge. We're good to go.
